onnxruntime/core/providers/cpu/tensor/unique.cc

```diff
@@ -0,0 +1,354 @@
+
+    for (int64_t row = 0; row < num_rows; ++row) {
+      // copy num_cols items from entries to output
+      if (std::is_same<T, std::string>::value) {
+        std::copy(item, item + num_cols, &Y_data[out_offset]);
+      } else {
+        std::copy_n(item, num_cols, &Y_data[out_offset]);
+      }
+
+      item += num_cols;
+      out_offset += num_unique * num_cols;
+    }
+
+    assert(item == items.cend());
+
+    if (indices_out) {
+      indices_data[output_idx] = indices[unsorted_idx].front();
+    }
+
+    if (counts) {
+      counts_data[output_idx] = indices[unsorted_idx].size();
+    }
+  }
+
+  if (inverse_indices) {
+    if (sorted) {
+      // need to convert unsorted entries in the inverse index to their sorted values
+      std::vector<int64_t> unsorted_to_sorted;
+      unsorted_to_sorted.resize(num_unique);
+      int64_t sorted_idx = 0;
+      for (const auto& offset : offsets) {
+        unsorted_to_sorted[offset.second] = sorted_idx++;
+      }
+
+      for (size_t i = 0, end = inverse_index.size(); i < end; ++i) {
+        inverse_indices_data[i] = unsorted_to_sorted[inverse_index[i]];
+      }
+    } else {
+      for (size_t i = 0, end = inverse_index.size(); i < end; ++i) {
+        inverse_indices_data[i] = inverse_index[i];
+      }
+    }
+  }
+}
+
+template <typename T>
+Status Unique::ComputeImpl(OpKernelContext& context) const {
+  const Tensor& input = *context.Input<Tensor>(0);
+  auto data = input.DataAsSpan<T>();
+
+  if (flatten_) {
+    std::map<const T, int64_t> offsets;  // offset of entry in indices. provides map between sorted and unsorted values
+    std::vector<std::vector<int64_t>> indices;
+    std::vector<int64_t> inverse_index;
+
+    indices.reserve(data.size() / 2);  // arbitrary value. at worst 1 realloc but could be too large
+    inverse_index.reserve(data.size());
+
+    int64_t num_unique = 0;
+
+    for (int64_t i = 0, end = input.Shape().Size(); i < end; ++i) {
+      auto entry = offsets.find(data[i]);
+      if (entry == offsets.end()) {
+        offsets[data[i]] = num_unique;
+        inverse_index.push_back({num_unique});
+        indices.push_back({i});
+        ++num_unique;
+      } else {
+        size_t indices_idx = entry->second;
+        indices[indices_idx].push_back(i);
+        inverse_index.push_back(indices_idx);
+      }
+    }
+
+    CreateFlattenedOutput(context, offsets, indices, inverse_index, sort_);
+  } else {
+    const auto& input_shape = input.Shape();
+    const int64_t input_dims = static_cast<int64_t>(input_shape.NumDimensions());
+    const int64_t axis = HandleNegativeAxis(axis_, input_dims);
+
+    std::vector<int64_t> subtensor_dims;
+    subtensor_dims.reserve(input_dims);
+    for (int64_t i = 0; i < input_dims; ++i) {
+      subtensor_dims.push_back(i == axis ? 1 : input_shape[i]);
+    }
+
+    TensorShape subtensor_shape(std::move(subtensor_dims));
+
+    std::map<const Subtensor<T>, int64_t> offsets;
```

Comment:
Would it help "average case" perf if we maintained an `unordered_map` instead of a `map` in case we knew the output need not be sorted ? I think the average case is going to only ask for unique elements instead of sorted unique elements. Models that don't require sorted uniques need not pay the cost...

Reply:
Default is for output to be sorted, so I assumed that was the most expected use case. If that isn't the case, the default shouldn't be sorted output. Side note: numpy.unique doesn't support unsorted output so that would seem like an additional signal that sorted is more important.

Would result in quite a bit of additional code to have an optimal path for the unsorted use case. I would suggest we defer this sort of performance optimization until we have a model that uses unsorted Unique in order to assess what the best balance is.

Reply:
Also may not be as simple as just using unordered_map. You need to preserve the original order, but resizing of unordered_map as items are added won't guarantee the order that things were originally added to it is maintained.

Reply:
that's true. that does complicate things.